### PR TITLE
customizable transform statistics

### PR DESCRIPTION
**Pull request by Zenglinxiao**
Currently, `TransformStatistics` only tracks statistics of providing Transforms, we need to modify this class if we want to track statistics from customized Transforms created by ourselves, which is not convenient.
In this PR, we made it possible to track statistics for customized transform.

### CODE REVIEW

**Comment by francoishernandez** [comment]
This is nice.
Maybe it would be good to reduce the verbosity of the stats logging also. I don't remember what's the exact logic here but it seems these stats are logged way too frequently. Maybe we should log those only at fixed interval e.g. like we perform validation at `valid_steps` interval.

**Comment by Zenglinxiao** [comment]
> Maybe it would be good to reduce the verbosity of the stats logging also. I don't remember what's the exact logic here but it seems these stats are logged way too frequently. Maybe we should log those only at fixed interval e.g. like we perform validation at `valid_steps` interval.

These transform statistics and corpora loading log are only logged one time when begin(for load) or finish(for statistic) iterating the corpora. The log seems too frequently because: the same corpus are loaded in every producer process and iterating with stride/offset mechanism, this makes the same corpus loading `world_size` times faster w.r.t single process as reduced size, and each time `world_size` time log because of multiprocessing, thus resulting `world_size` * `world_size` times frequent logging.

I think we can disable other processes' logging and only let the first producer do the log.

**Comment by francoishernandez** [comment]
If we let only the first producer do the logging, we should have some way to aggregate the stats over all the processes, like the `all_gather_stats` of `onmt.utils.Statistics`.

File: onmt/inputters/corpus.py [hunk 1]
```diff
@@ -229,7 +232,9 @@ def _add_index(self, stream):
 
     def _iter_corpus(self):
         corpus_stream = self.corpus.load(
-            stride=self.stride, offset=self.offset)
+            stride=self.stride, offset=self.offset,
+            log_level=self.skip_empty_level
```
Comment (francoishernandez) [inline onmt/inputters/corpus.py:236]: not very straightforward to use `skip_empty_level` here...
maybe rename this flag to `data_loading_level` (and retain backcompatibility by keeping `skip_empty_level` but assigning its value to `data_loading_level` + warn for deprecation)

File: onmt/utils/distributed.py [hunk 2]
```diff
@@ -158,9 +158,10 @@ def signal_handler(self, signalnum, stackframe):
         raise Exception(msg)
 
 
-def batch_producer(generator_to_serve, queue, semaphore, opt):
+def batch_producer(generator_to_serve, queue, semaphore, opt, device_id):
     """Produce batches to `queues` from `generator_to_serve`."""
-    init_logger(opt.log_file)
+    if opt.verbose or device_id == 0:
```
Comment (francoishernandez) [inline onmt/utils/distributed.py:163]: What happens if an error is raised in other processes?

**Comment by Zenglinxiao** [comment]
Gather stats across all producer process is a little bit complex since it requires inter-process communication or shared object and needs extra effort which may introduce additional complexity while not adding too many benefits.
Here, I add the option `--verbose` to the `train.py` script, making it log only the first producer by default, and can get the full log if we set `--verbose`.
And `--skip_empty_level` option is also used here to default log the loading corpus name rather than its path. We can get the full corpus path if set `--skip_empty_level=error`, or completely silent loading corpus if set `--skip_empty_level=silent`. Since the use case is exactly the same as `skip_empty_level`, I reuse this option rather than create another option.

**Review by francoishernandez — COMMENTED** [review]
A few comments.
Also, it could be nice to add some transforms to one of the integration tests, to allow a quick look at how stats logging looks like.

File: onmt/utils/distributed.py (hunk 2 above)
Comment (Zenglinxiao) [reply onmt/utils/distributed.py:163]: So maybe we set the log level to be error/warning for other processes?

File: onmt/inputters/corpus.py [hunk 3]
```diff
@@ -123,7 +123,10 @@ def load(self, offset=0, stride=1):
         with exfile_open(self.src, mode='rb') as fs,\
                 exfile_open(self.tgt, mode='rb') as ft,\
                 exfile_open(self.align, mode='rb') as fa:
-            logger.info(f"Loading {repr(self)}...")
+            if log_level == "error":
+                logger.info(f"Loading {str(self)}...")
+            elif log_level == "warning":
+                logger.info(f"Loading {self.id}...")
```
[the remark this answers is not in the document]
Comment (Zenglinxiao) [reply onmt/inputters/corpus.py:129]: The `warning` `error` is adapted from `--skip_empty_level`. But here we use it to indicate the brevity of logging. so `logger.info` is what I meant for.

**Comment by Zenglinxiao** [comment]
Largely reduced Data related loggings by:
1. Assign corpus loading info log to `WeightedMixer`, and only log loaded times so far for each corpus when a new load is performed;
2. Logs related to data default to be reported only by the first producer process to reduce the logging amount, but can be reactivated to all producers if set `--verbose`.

**Pull request merged**